flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/enumerate/NonSplittingRegexEnumerator.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.file.src.enumerate;
+
+import org.apache.flink.connector.file.src.FileSourceSplit;
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.function.Predicate;
+
+/**
+ * This {@code FileEnumerator} enumerates all files under the given paths recursively. Each file
+ * matched the given regex pattern becomes one split; this enumerator does not split files into
+ * smaller "block" units.
+ *
+ * <p>The default instantiation of this enumerator filters files with the common hidden file
+ * prefixes '.' and '_'. A custom file filter can be specified.
+ */
+public class NonSplittingRegexEnumerator extends NonSplittingRecursiveEnumerator {
+
+    /** The custom filter predicate to filter out unwanted files. */
+    private final Predicate<Path> fileFilter;
+
+    private final RegexFileFilter regexFileFilter;
```

Comment:
Do we really need this in here? I think in here `fileFilter` is enough. We can always construnct the `fileFilter` using regex expression.

Reply:
I mean construct the `fileFilter`  in `FileSystemTableSource`.

Reply:
Also, I'm now thing do we really need to introduce `NonSplittingRecursiveEnumerator`?
Why don't reuse `NonSplittingRecursiveEnumerator`, but the `fileFilter` will compose the logic of `DefaultFileFilter` and `RegexFileFilter`.

Reply:
Because it does not satisfy some situations. 
For example, we have files as follows.
```
/
 |- dir
     |- test
           |- test.java
           |- innerDir
                 |- test.java
```
We want to read these two `test.java` under the directory `/dir`. Then we will use the final regex `/dir/.*/test.java`.
If we reuse `NonSplittingRecursiveEnumerator ` + `fileFilter`, it will skip the directory `/dir/test` because its path does not match the regex `/dir/.*/test.java`. And it will lose the files.

Reply:
I think it is not suitable when the regex is complex. I think I should remove the `connectBasePathAndRegex ` and use the actual value passed by `source.regex-pattern`.

But I still think we should keep the table path as the base dir to start to search files.